common-rest/src/main/java/org/uniprot/api/rest/download/queue/BasicAbstractMessageListener.java

```diff
@@ -0,0 +1,261 @@
+            RabbitTemplate rabbitTemplate) {
+        this.downloadConfigProperties = downloadConfigProperties;
+        this.asyncDownloadQueueConfigProperties = asyncDownloadQueueConfigProperties;
+        this.jobRepository = jobRepository;
+        this.rabbitTemplate = rabbitTemplate;
+    }
+
+    @Override
+    public void onMessage(Message message) {
+        String jobId = null;
+        DownloadJob downloadJob = null;
+        try {
+            jobId = message.getMessageProperties().getHeader(JOB_ID_HEADER);
+            log.info("Received job {} in listener", jobId);
+            if (isMaxRetriedReached(message)) {
+                sendToUndeliveredQueue(jobId, message);
+                log.error(
+                        "Message with job id {} discarded after max retry {}",
+                        jobId,
+                        getMaxRetryCount());
+                return;
+            }
+
+            Optional<DownloadJob> optDownloadJob = this.jobRepository.findById(jobId);
+            String errorMsg = "Unable to find jobId " + jobId + " in db";
+            downloadJob = optDownloadJob.orElseThrow(() -> new MessageListenerException(errorMsg));
+
+            processMessage(message, downloadJob);
+            // get the fresh object from redis
+            optDownloadJob = this.jobRepository.findById(jobId);
+            if (optDownloadJob.isPresent()) {
+                downloadJob = optDownloadJob.get();
+                if (downloadJob.getStatus() == JobStatus.FINISHED) {
+                    log.info("Message with jobId {} processed successfully", jobId);
+                }
+            }
+        } catch (Exception ex) {
+            if (getRetryCountByBroker(message) <= getMaxRetryCount()) {
+                log.error("Download job id {} failed with error {}", jobId, ex.getStackTrace());
+                Message updatedMessage = addAdditionalHeaders(message, ex);
+                updateDownloadJob(updatedMessage, downloadJob, JobStatus.ERROR);
+                sendToRetryQueue(jobId, updatedMessage);
+            } else {
+                // the flow should not come here, letting the flow complete without rethrowing
+                // exception to avoid poison message
+                log.error(
+                        "Message with jobId {} failed due to error {} which is not handled",
+                        jobId,
+                        ex);
+            }
+        }
+    }
+
+    protected abstract void processMessage(Message message, DownloadJob downloadJob);
+
+    protected static boolean isJobSeenBefore(
+            DownloadJob downloadJob, Path idsFile, Path resultFile) {
+        return Files.exists(idsFile)
+                && Files.exists(resultFile)
+                && downloadJob.getStatus() != JobStatus.ERROR;
+    }
+
+    protected void saveIdsInTempFile(Path filePath, Stream<String> ids) throws IOException {
+        Iterable<String> source = ids::iterator;
```

Comment:
I don't remember why did we choose this name saveIdsInTempFile. We could use better name 

1. saveIdentifiersToFileSystem
2. writeIdentifiersToFile

Reply:
writeIdsToFileSystem